crates/api/src/frame_info.rs

```diff
@@ -0,0 +1,180 @@
+    /// This is a global cache of backtrace frame information for all active
+    ///
+    /// This global cache is used during `Trap` creation to symbolicate frames.
+    /// This is populated on module compilation, and it is cleared out whenever
+    /// all references to a module are dropped.
+    pub static ref FRAME_INFO: GlobalFrameInfo = GlobalFrameInfo::default();
+}
+
+#[derive(Default)]
+pub struct GlobalFrameInfo {
+    /// An internal map that keeps track of backtrace frame information for
+    /// each module.
+    ///
+    /// This map is morally a map of ranges to a map of information for that
+    /// module. Each module is expected to reside in a disjoint section of
+    /// contiguous memory. No modules can overlap.
+    ///
+    /// The key of this map is the highest address in the module and the value
+    /// is the module's information, which also contains the start address.
+    ranges: RwLock<BTreeMap<usize, ModuleFrameInfo>>,
+}
+
+/// An RAII structure used to unregister a module's frame information when the
+/// module is destroyed.
+pub struct GlobalFrameInfoRegistration {
+    /// The key that will be removed from the global `ranges` map when this is
+    /// dropped.
+    key: usize,
+}
+
+struct ModuleFrameInfo {
+    start: usize,
+    functions: BTreeMap<usize, (usize, FuncIndex)>,
+    names: Arc<Names>,
+}
+
+impl GlobalFrameInfo {
+    /// Registers a new compiled module's frame information.
+    ///
+    /// This function will register the `names` information for all of the
+    /// compiled functions within `module`. If the `module` has no functions
+    /// then `None` will be returned. Otherwise the returned object, when
+    /// dropped, will be used to unregister all name information from this map.
+    pub fn register(
+        &self,
+        names: &Arc<Names>,
+        module: &CompiledModule,
+    ) -> Option<GlobalFrameInfoRegistration> {
+        let mut min = usize::max_value();
+        let mut max = 0;
+        let mut functions = BTreeMap::new();
+        for (i, allocated) in module.finished_functions() {
+            let (start, end) = unsafe {
+                let ptr = (**allocated).as_ptr();
+                let len = (**allocated).len();
+                (ptr as usize, ptr as usize + len)
+            };
+            if start < min {
+                min = start;
+            }
+            if end > max {
+                max = end;
+            }
+            let func_index = module.module().func_index(i);
+            assert!(functions.insert(end, (start, func_index)).is_none());
+        }
+        if functions.len() == 0 {
+            return None;
+        }
+
+        let mut ranges = self.ranges.write().unwrap();
+        if let Some((_, prev)) = ranges.range(max..).next() {
+            assert!(prev.start > max);
+        }
+        if let Some((prev_end, _)) = ranges.range(..=min).next_back() {
+            assert!(*prev_end < min);
+        }
+        assert!(ranges
```

Comment:
I was lost in these assert-s for a sec. Can we re-arrange the logic to have main stuff outside of assert!, and asserting one inside? e.g. `assert!(ranges.range(max..).next().map_or(true, |(_,prev)| ....` and `let _inserted = ranges.insert(..); assert!(_inserted.is_none());`?